File: typescript/src/modules/player.ts
/**
 * 玩家管理服务模块
 * 基于 @sttot/axios-api 的标准化API定义
 */

import { apiBase } from '@sttot/axios-api';
import {
  Player,
  CreatePlayerRequest,
  GetPlayerRequest,
  UpdatePlayerRequest,
  DeletePlayerRequest,
  BanPlayerRequest,
  UnbanPlayerRequest,
  ListPlayersRequest,
  ListPlayersResponse,
  ValidateRequest,
  ValidateResponse,
  EmptyResponse
} from '../types';
import { commonErrorHandler } from '../utils/errorHandler';

export const initPlayerService = (apiFactory: ReturnType<typeof apiBase>) => {
  class PlayerService {
    // 创建玩家
    public createPlayer = apiFactory<CreatePlayerRequest, Player>(
      // 生成 RequestConfig
      (request) => ({
        method: 'POST',
        url: '/api/v1/players',
        data: {
          name: request.name,
          town_id: request.townId,
          qq: request.qq,
          qqguild: request.qqguild,
          discord: request.discord,
          in_qq_group: request.inQqGroup,
          in_qq_guild: request.inQqGuild,
          in_discord: request.inDiscord
        }
      }),
      // 处理响应：转换snake_case到camelCase
      ({ data }) => ({
        id: data.id,
        name: data.name,
        townId: data.town_id,
        qq: data.qq,
        qqguild: data.qqguild,
        discord: data.discord,
        inQqGroup: data.in_qq_group,
        inQqGuild: data.in_qq_guild,
        inDiscord: data.in_discord,
        banMode: data.ban_mode,
        banExpire: data.ban_expire,
        banReason: data.ban_reason,
        createdAt: data.created_at,
        updatedAt: data.updated_at
      }),
      // 处理错误：解析新的错误格式 {"detail": "..."}
      commonErrorHandler
    );

    // 获取玩家列表
    public listPlayers = apiFactory<ListPlayersRequest, ListPlayersResponse>(
      (request = {}) => ({
        method: 'GET',
        url: '/api/v1/players',
        params: this.buildParams({
          page: request.page || 1,
          page_size: request.pageSize || 20,
          search: request.search,
          town_id: request.townId,
          ban_mode: request.banMode,
          name: request.name,
          qq: request.qq,
          qqguild: request.qqguild,
          discord: request.discord
        })
      }),
      ({ data }) => ({
        players: data.players.map((player: any) => ({
          id: player.id,
          name: player.name,
          townId: player.town_id,
          qq: player.qq,
          qqguild: player.qqguild,
          discord: player.discord,
          inQqGroup: player.in_qq_group,
          inQqGuild: player.in_qq_guild,
          inDiscord: player.in_discord,
          banMode: player.ban_mode,
          banExpire: player.ban_expire,
          banReason: player.ban_reason,
          createdAt: player.created_at,
          updatedAt: player.updated_at
        })),
        total: data.total,
        page: data.page,
        pageSize: data.page_size
      }),
      commonErrorHandler
    );

    // 获取玩家详情
    public getPlayer = apiFactory<GetPlayerRequest, Player>(
      (request) => ({
        method: 'GET',
        url: `/api/v1/players/${request.id}`
      }),
      ({ data }) => ({
        id: data.id,
        name: data.name,
        townId: data.town_id,
        qq: data.qq,
        qqguild: data.qqguild,
        discord: data.discord,
        inQqGroup: data.in_qq_group,
        inQqGuild: data.in_qq_guild,
        inDiscord: data.in_discord,
        banMode: data.ban_mode,
        banExpire: data.ban_expire,
        banReason: data.ban_reason,
        createdAt: data.created_at,
        updatedAt: data.updated_at
      }),
      commonErrorHandler
    );

    // 更新玩家信息
    public updatePlayer = apiFactory<UpdatePlayerRequest, Player>(
      (request) => ({
        method: 'PUT',
        url: `/api/v1/players/${request.id}`,
        data: {
          name: request.name,
          town_id: request.townId,
          qq: request.qq,
          qqguild: request.qqguild,
          discord: request.discord,
          in_qq_group: request.inQqGroup,
          in_qq_guild: request.inQqGuild,
          in_discord: request.inDiscord
        }
      }),
      ({ data }) => ({
        id: data.id,
        name: data.name,
        townId: data.town_id,
        qq: data.qq,
        qqguild: data.qqguild,
        discord: data.discord,
        inQqGroup: data.in_qq_group,
        inQqGuild: data.in_qq_guild,
        inDiscord: data.in_discord,
        banMode: data.ban_mode,
        banExpire: data.ban_expire,
        banReason: data.ban_reason,
        createdAt: data.created_at,
        updatedAt: data.updated_at
      }),
      commonErrorHandler
    );

    // 删除玩家
    public deletePlayer = apiFactory<DeletePlayerRequest, EmptyResponse>(
      (request) => ({
        method: 'DELETE',
        url: `/api/v1/players/${request.id}`
      }),
      ({ data }) => data as EmptyResponse,
      commonErrorHandler
    );

    // 封禁玩家
    public banPlayer = apiFactory<BanPlayerRequest, EmptyResponse>(
      (request) => ({
        method: 'POST',
        url: `/api/v1/players/${request.playerId}/ban`,
        data: {
          ban_mode: request.banMode,
          duration_seconds: request.durationSeconds,
          reason: request.reason
        }
      }),
      ({ data }) => data as EmptyResponse,
      commonErrorHandler
    );

    // 解封玩家
    public unbanPlayer = apiFactory<UnbanPlayerRequest, EmptyResponse>(
      (request) => ({
        method: 'POST',
        url: `/api/v1/players/${request.playerId}/unban`,
        data: {}
      }),
      ({ data }) => data as EmptyResponse,
      commonErrorHandler
    );

    // 批量玩家验证
    public validate = apiFactory<ValidateRequest, ValidateResponse>(
      (request) => ({
        method: 'POST',
        url: '/api/v1/players/validate',
        data: {
          players: request.players.map(p => ({
            player_name: p.playerName,
            ip: p.ip,
            client_version: p.clientVersion,
            protocol_version: p.protocolVersion
          })),
          server_id: request.serverId,
          login: request.login
        }
      }),
      ({ data }) => ({
        results: data.results.map((result: any) => ({
          playerName: result.player_name,
          allowed: result.allowed,
          playerId: result.player_id,
          reason: result.reason,
          newbie: result.newbie,
          banMode: result.ban_mode,
          banExpire: result.ban_expire,
          banReason: result.ban_reason
        })),
        processedAt: data.processed_at
      }),
      commonErrorHandler
    );

    // 辅助方法：构建查询参数，过滤掉undefined值
    public buildParams(params: Record<string, any>): Record<string, any> {
      const result: Record<string, any> = {};
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined && value !== null) {
          result[key] = value;
        }
      }
      return result;
    }
  }

  return new PlayerService();
};

File: golang/modules/player.go
package modules

import (
	"context"
	"strconv"

	"github.com/NewNanCity/NewNanManager-SDK/clients/golang/utils"
	"github.com/go-resty/resty/v2"
)

// PlayerService 玩家管理服务
type PlayerService struct {
	client *resty.Client
}

// NewPlayerService 创建玩家服务实例
func NewPlayerService(client *resty.Client) *PlayerService {
	return &PlayerService{client: client}
}

// ListPlayers 获取玩家列表
func (s *PlayerService) ListPlayers(page, pageSize *int32, search *string, townID *int32, banMode *BanMode, name, qq, qqguild, discord *string) (*PlayersListData, error) {
	return s.ListPlayersWithContext(context.Background(), page, pageSize, search, townID, banMode, name, qq, qqguild, discord)
}

// ListPlayersWithContext executes ListPlayers with cancellation scoped to this request.
func (s *PlayerService) ListPlayersWithContext(ctx context.Context, page, pageSize *int32, search *string, townID *int32, banMode *BanMode, name, qq, qqguild, discord *string) (*PlayersListData, error) {
	req := s.client.R().SetContext(ctx)

	if page != nil {
		req.SetQueryParam("page", strconv.Itoa(int(*page)))
	}
	if pageSize != nil {
		req.SetQueryParam("page_size", strconv.Itoa(int(*pageSize)))
	}
	if search != nil {
		req.SetQueryParam("search", *search)
	}
	if townID != nil {
		req.SetQueryParam("town_id", strconv.Itoa(int(*townID)))
	}
	if banMode != nil {
		req.SetQueryParam("ban_mode", strconv.Itoa(int(*banMode)))
	}
	if name != nil {
		req.SetQueryParam("name", *name)
	}
	if qq != nil {
		req.SetQueryParam("qq", *qq)
	}
	if qqguild != nil {
		req.SetQueryParam("qqguild", *qqguild)
	}
	if discord != nil {
		req.SetQueryParam("discord", *discord)
	}

	resp, err := req.Get("/api/v1/players")

	var result PlayersListData
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// CreatePlayer 创建玩家
func (s *PlayerService) CreatePlayer(request CreatePlayerRequest) (*Player, error) {
	return s.CreatePlayerWithContext(context.Background(), request)
}

// CreatePlayerWithContext executes CreatePlayer with cancellation scoped to this request.
func (s *PlayerService) CreatePlayerWithContext(ctx context.Context, request CreatePlayerRequest) (*Player, error) {
	resp, err := s.client.R().SetContext(ctx).
		SetBody(request).
		Post("/api/v1/players")

	var result Player
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// Validate 玩家验证（支持批处理）
func (s *PlayerService) Validate(request ValidateRequest) (*ValidateResponse, error) {
	return s.ValidateWithContext(context.Background(), request)
}

// ValidateWithContext executes Validate with cancellation scoped to this request.
func (s *PlayerService) ValidateWithContext(ctx context.Context, request ValidateRequest) (*ValidateResponse, error) {
	return s.validateWithContext(ctx, request, nil)
}

// ValidateWithSession sends the server session fencing headers.
func (s *PlayerService) ValidateWithSession(request ValidateRequest, session SessionContext) (*ValidateResponse, error) {
	return s.ValidateWithContextAndSession(context.Background(), request, session)
}

// ValidateWithContextAndSession scopes cancellation and sends session fencing headers.
func (s *PlayerService) ValidateWithContextAndSession(ctx context.Context, request ValidateRequest, session SessionContext) (*ValidateResponse, error) {
	return s.validateWithContext(ctx, request, &session)
}

func (s *PlayerService) validateWithContext(ctx context.Context, request ValidateRequest, session *SessionContext) (*ValidateResponse, error) {
	req := s.client.R().SetContext(ctx)
	if session != nil {
		if err := applySessionHeaders(req, *session); err != nil {
			return nil, err
		}
	}
	resp, err := req.
		SetBody(request).
		Post("/api/v1/players/validate")

	var result ValidateResponse
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// GetPlayer 获取玩家详情
func (s *PlayerService) GetPlayer(id int32) (*Player, error) {
	return s.GetPlayerWithContext(context.Background(), id)
}

// GetPlayerWithContext executes GetPlayer with cancellation scoped to this request.
func (s *PlayerService) GetPlayerWithContext(ctx context.Context, id int32) (*Player, error) {
	resp, err := s.client.R().SetContext(ctx).
		Get("/api/v1/players/" + strconv.Itoa(int(id)))

	var result Player
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// UpdatePlayer 更新玩家信息
func (s *PlayerService) UpdatePlayer(id int32, request UpdatePlayerRequest) (*Player, error) {
	return s.UpdatePlayerWithContext(context.Background(), id, request)
}

// UpdatePlayerWithContext executes UpdatePlayer with cancellation scoped to this request.
func (s *PlayerService) UpdatePlayerWithContext(ctx context.Context, id int32, request UpdatePlayerRequest) (*Player, error) {
	resp, err := s.client.R().SetContext(ctx).
		SetBody(request).
		Put("/api/v1/players/" + strconv.Itoa(int(id)))

	var result Player
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// DeletePlayer 删除玩家
func (s *PlayerService) DeletePlayer(id int32) error {
	return s.DeletePlayerWithContext(context.Background(), id)
}

// DeletePlayerWithContext executes DeletePlayer with cancellation scoped to this request.
func (s *PlayerService) DeletePlayerWithContext(ctx context.Context, id int32) error {
	resp, err := s.client.R().SetContext(ctx).
		Delete("/api/v1/players/" + strconv.Itoa(int(id)))

	return utils.HandleResponse(resp, err, nil)
}

// BanPlayer 封禁玩家
func (s *PlayerService) BanPlayer(playerID int32, request BanPlayerRequest) error {
	return s.BanPlayerWithContext(context.Background(), playerID, request)
}

// BanPlayerWithContext executes BanPlayer with cancellation scoped to this request.
func (s *PlayerService) BanPlayerWithContext(ctx context.Context, playerID int32, request BanPlayerRequest) error {
	resp, err := s.client.R().SetContext(ctx).
		SetBody(request).
		Post("/api/v1/players/" + strconv.Itoa(int(playerID)) + "/ban")

	return utils.HandleResponse(resp, err, nil)
}

// UnbanPlayer 解封玩家
func (s *PlayerService) UnbanPlayer(playerID int32) error {
	return s.UnbanPlayerWithContext(context.Background(), playerID)
}

// UnbanPlayerWithContext executes UnbanPlayer with cancellation scoped to this request.
func (s *PlayerService) UnbanPlayerWithContext(ctx context.Context, playerID int32) error {
	resp, err := s.client.R().SetContext(ctx).
		Post("/api/v1/players/" + strconv.Itoa(int(playerID)) + "/unban")

	return utils.HandleResponse(resp, err, nil)
}
